File: src/server/kube/discovery.ts
import type { KubeConfig } from '@kubernetes/client-node'
import { KubeManager } from './manager'
import type { ResourceDefinition, ResourceGroup } from './types'

const WORKLOAD_KINDS = new Set([
  'Pod',
  'Deployment',
  'ReplicaSet',
  'StatefulSet',
  'DaemonSet',
  'Job',
  'CronJob',
  'ReplicationController',
])

const NETWORKING_KINDS = new Set([
  'Service',
  'Ingress',
  'Endpoints',
  'NetworkPolicy',
  'EndpointSlice',
])

const STORAGE_KINDS = new Set([
  'PersistentVolumeClaim',
  'PersistentVolume',
  'StorageClass',
])

const CONFIG_KINDS = new Set([
  'ConfigMap',
  'Secret',
  'ServiceAccount',
  'ResourceQuota',
  'LimitRange',
])

function getGroupLabel(resource: ResourceDefinition): string {
  if (WORKLOAD_KINDS.has(resource.kind)) return 'Workloads'
  if (NETWORKING_KINDS.has(resource.kind)) return 'Networking'
  if (STORAGE_KINDS.has(resource.kind)) return 'Storage'
  if (CONFIG_KINDS.has(resource.kind)) return 'Configuration'
  if (resource.group && !resource.group.endsWith('.k8s.io')) return 'Custom Resources'
  return 'Other'
}

export function groupResources(resources: ResourceDefinition[]): ResourceGroup[] {
  const groups = new Map<string, ResourceDefinition[]>()

  for (const resource of resources) {
    const label = getGroupLabel(resource)
    const existing = groups.get(label)
    if (existing) {
      existing.push(resource)
    } else {
      groups.set(label, [resource])
    }
  }

  const order = ['Workloads', 'Networking', 'Storage', 'Configuration', 'Custom Resources', 'Other']
  return order
    .filter((label) => groups.has(label))
    .map((label) => ({ label, resources: groups.get(label)! }))
}

type APIResourceResponse = {
  groupVersion: string
  resources: Array<{
    name: string
    singularName: string
    namespaced: boolean
    kind: string
    verbs: string[]
    shortNames?: string[]
    categories?: string[]
  }>
}

async function fetchJSON<T>(kubeConfig: KubeConfig, path: string): Promise<T> {
  const cluster = kubeConfig.getCurrentCluster()
  if (!cluster) {
    throw new Error('No active cluster found in kubeconfig')
  }
  const url = `${cluster.server}${path}`
  const fetchOpts = await kubeConfig.applyToFetchOptions({} as Parameters<typeof kubeConfig.applyToFetchOptions>[0])
  const response = await fetch(url, {
    ...fetchOpts,
    headers: {
      ...(fetchOpts.headers as Record<string, string> | undefined),
      Accept: 'application/json',
    },
  })
  if (!response.ok) {
    throw new Error(`HTTP ${response.status}: ${response.statusText}`)
  }
  return response.json() as Promise<T>
}

function parseResources(
  data: APIResourceResponse,
  group: string,
  version: string,
): ResourceDefinition[] {
  return data.resources
    .filter((r) => !r.name.includes('/'))
    .map((r) => ({
      group,
      version,
      kind: r.kind,
      name: r.name,
      singularName: r.singularName,
      namespaced: r.namespaced,
      verbs: r.verbs,
      shortNames: r.shortNames ?? [],
      categories: r.categories ?? [],
    }))
}

type APIGroupListResponse = {
  groups: Array<{
    name: string
    preferredVersion: {
      groupVersion: string
      version: string
    }
  }>
}

export async function discoverAPIs(): Promise<ResourceDefinition[]> {
  const manager = KubeManager.getInstance()
  const kubeConfig = manager.getKubeConfig()
  const resources: ResourceDefinition[] = []

  const coreData = await fetchJSON<APIResourceResponse>(kubeConfig, '/api/v1')
  resources.push(...parseResources(coreData, '', 'v1'))

  const groupsData = await fetchJSON<APIGroupListResponse>(kubeConfig, '/apis')
  const groupFetches = groupsData.groups.map(async (g) => {
    const gv = g.preferredVersion.groupVersion
    const version = g.preferredVersion.version
    try {
      const data = await fetchJSON<APIResourceResponse>(kubeConfig, `/apis/${gv}`)
      return parseResources(data, g.name, version)
    } catch {
      return []
    }
  })

  const groupResults = await Promise.all(groupFetches)
  for (const result of groupResults) {
    resources.push(...result)
  }

  return resources
}

File: backend/app/kube/discovery.py
import asyncio
from typing import Any

from app.kube.manager import KubeManager
from app.kube.resources import _kube_request
from app.models import ResourceDefinition, ResourceGroup

WORKLOAD_KINDS = frozenset([
    "Pod", "Deployment", "ReplicaSet", "StatefulSet",
    "DaemonSet", "Job", "CronJob", "ReplicationController",
])

NETWORKING_KINDS = frozenset([
    "Service", "Ingress", "Endpoints", "NetworkPolicy", "EndpointSlice",
])

STORAGE_KINDS = frozenset([
    "PersistentVolumeClaim", "PersistentVolume", "StorageClass",
])

CONFIG_KINDS = frozenset([
    "ConfigMap", "Secret", "ServiceAccount", "ResourceQuota", "LimitRange",
])

GROUP_ORDER = ["Workloads", "Networking", "Storage", "Config", "Custom Resources", "Other"]


def _get_group_label(resource: ResourceDefinition) -> str:
    if resource.kind in WORKLOAD_KINDS:
        return "Workloads"
    if resource.kind in NETWORKING_KINDS:
        return "Networking"
    if resource.kind in STORAGE_KINDS:
        return "Storage"
    if resource.kind in CONFIG_KINDS:
        return "Config"
    if resource.group and not resource.group.endswith(".k8s.io"):
        return "Custom Resources"
    return "Other"


def group_resources(resources: list[ResourceDefinition]) -> list[ResourceGroup]:
    groups: dict[str, list[ResourceDefinition]] = {}
    for resource in resources:
        label = _get_group_label(resource)
        groups.setdefault(label, []).append(resource)
    return [
        ResourceGroup(label=label, resources=groups[label])
        for label in GROUP_ORDER
        if label in groups
    ]


def _parse_resources(
    data: dict[str, Any], group: str, version: str
) -> list[ResourceDefinition]:
    result: list[ResourceDefinition] = []
    for r in data.get("resources", []):
        if "/" in r["name"]:
            continue
        result.append(
            ResourceDefinition(
                group=group,
                version=version,
                kind=r["kind"],
                name=r["name"],
                singular_name=r.get("singularName", ""),
                namespaced=r["namespaced"],
                verbs=r.get("verbs", []),
                short_names=r.get("shortNames", []),
                categories=r.get("categories", []),
            )
        )
    return result


async def _fetch_group_resources(
    group_name: str, group_version: str, version: str
) -> list[ResourceDefinition]:
    api_client = KubeManager.get_instance().get_api_client()
    try:
        data = await asyncio.to_thread(
            _kube_request, api_client, f"/apis/{group_version}"
        )
        return _parse_resources(data, group_name, version)
    except Exception:
        return []


async def discover_apis() -> list[ResourceDefinition]:
    api_client = KubeManager.get_instance().get_api_client()

    core_data = await asyncio.to_thread(_kube_request, api_client, "/api/v1")
    resources = _parse_resources(core_data, "", "v1")

    groups_data = await asyncio.to_thread(_kube_request, api_client, "/apis")
    tasks = [
        _fetch_group_resources(
            g["name"],
            g["preferredVersion"]["groupVersion"],
            g["preferredVersion"]["version"],
        )
        for g in groups_data.get("groups", [])
    ]

    group_results = await asyncio.gather(*tasks)
    for result in group_results:
        resources.extend(result)

    return resources
